### PR TITLE
Extend Aitken acceleration to handle multiple data sets

### LINKED ISSUE #1184: Assertion instead of Error Message when using Aitken in "coupling-scheme:multi"
https://github.com/precice/precice/blob/bf3b2cc31619152ab6d2c702777eaf95eb7c53ba/src/acceleration/AitkenAcceleration.cpp#L42

I understand I am not supposed to use `Aitken` acceleration in parallel implicit schemes but anyway it should give me an error here rather it is giving me an assertion:
```
ASSERTION FAILED
Location:   virtual void precice::acceleration::AitkenAcceleration::initialize(precice::acceleration::Acceleration::DataMap&)
File:       /home/mansari/precice-2.3.0/src/acceleration/AitkenAcceleration.cpp:42
Expression: _dataIDs.size() == 2
Rank:       0
Arguments:  none
```

Following is my part of config where I am defining coupling scheme:

```xml
<coupling-scheme:multi>
      <participant name="Fluid" control="yes" />
      <participant name="Calculix" />
      <participant name="NutilsRight" />
      <max-time-windows value="200" />
      <time-window-size value="1e-4" />
      <exchange data="Velocity" mesh="NutilsRight-Mesh" to="Fluid" from="NutilsRight"/>
      <exchange data="Pressure" mesh="FluidOutlet-Mesh" to="NutilsRight" from="Fluid"/>
      <exchange data="Force" mesh="Fluid-Mesh-Faces" from="Fluid" to="Calculix" />
      <exchange data="Displacement" mesh="Calculix_Mesh" from="Calculix" to="Fluid" />
      <max-iterations value="30" />
      <relative-convergence-measure limit="1.0e-3" data="Pressure" mesh="FluidOutlet-Mesh"/>
      <relative-convergence-measure limit="1.0e-3" data="Velocity" mesh="NutilsRight-Mesh"/>
      <relative-convergence-measure limit="1e-3" data="Displacement" mesh="Calculix_Mesh" />
      <relative-convergence-measure limit="1e-3" data="Force" mesh="Fluid-Mesh-Faces" />
      <acceleration:aitken>
        <data name="Displacement" mesh="Calculix_Mesh" />
        <!-- <data name="Force" mesh="Fluid-Mesh-Faces" /> -->
        <data name="Velocity" mesh="NutilsRight-Mesh" />
        <data name="Pressure" mesh="FluidOutlet-Mesh" />
        <initial-relaxation value="0.1" />
      </acceleration:aitken>
    </coupling-scheme:multi>
```

### LINKED ISSUE #1360: Uninformative assertion hit while using dynamic Aitken acceleration
**Describe your setup**

Operating system (e.g. Linux distribution and version): _Ubuntu 20.04_
preCICE Version: 2.4.0 commit: https://github.com/precice/precice/commit/67b0e6d9c9629b69d923cc540056bfd63a258a56

**Describe the problem**

While using dynamic Aitken acceleration scheme in running a case from the repository [coupled-heat-conduction](https://github.com/IshaanDesai/coupled-heat-conduction), the following assertion was hit:

```bash
[ipvs-epyc1:878755] [ 2] ASSERTION FAILED
Location:   virtual void precice::acceleration::AitkenAcceleration::initialize(const DataMap&)
File:       /home/desaiin/precice/src/acceleration/AitkenAcceleration.cpp:42
Expression: _dataIDs.size() == 2
Rank:       16
Arguments:  none
Stacktrace:
 0# getStacktrace[abi:cxx11]() in /home/desaiin/precice/build/lib/libprecice.so.2
 1# precice::acceleration::AitkenAcceleration::initialize(std::map<int, std::shared_ptr<precice::cplscheme::CouplingData>, std::less<int>, std::allocator<std::pair<int const, std::shared_ptr<precice::cplscheme::CouplingData> > > > const&) in /home/desaiin/precice/build/lib/libprecice.so.2
 2# precice::cplscheme::BaseCouplingScheme::initializeStorages() in /home/desaiin/precice/build/lib/libprecice.so.2
 3# precice::cplscheme::BaseCouplingScheme::initialize(double, int) in /home/desaiin/precice/build/lib/libprecice.so.2
 4# precice::impl::SolverInterfaceImpl::initialize() in /home/desaiin/precice/build/lib/libprecice.so.2
 5# precice::SolverInterface::initialize() in /home/desaiin/precice/build/lib/libprecice.so.2
 6# 0x00007F143A139A7D in /home/desaiin/.local/lib/python3.8/site-packages/cyprecice.cpython-38-x86_64-linux-gnu.so
```

The corresponding dynamic Aitken acceleration configuration part from the [config file](https://github.com/IshaanDesai/coupled-heat-conduction/blob/36ca2f53e6c75e499f03bd648c22163ef27d1593/precice-config.xml) is:

```xml
       <acceleration:aitken>
          <data name="porosity" mesh="macro-mesh"/>
          <data name="k_00" mesh="macro-mesh"/>
          <data name="k_11" mesh="macro-mesh"/>
          <initial-relaxation value="0.5"/>
       </acceleration:aitken>
```

**Step To Reproduce**

The assertion is hit while using dynamic Aitken acceleration scheme in the [macro-micro coupled case example](https://github.com/IshaanDesai/coupled-heat-conduction/tree/36ca2f53e6c75e499f03bd648c22163ef27d1593) from the repository [coupled-heat-conduction](https://github.com/IshaanDesai/coupled-heat-conduction). To run this example, one needs to install pyprecice and the micro-manager.

**Expected behaviour**

It is not clear to me why the dynamic Aitken acceleration would fail, hence the expected behavior would be that the case runs as usual and the implicit coupling is computed.

**Additional context**
[log_micro_simulations.log](https://github.com/precice/precice/files/9055615/log_micro_simulations.log)

### PR DESCRIPTION
**Pull request by davidscn**
## Main changes of this PR
Extends the Aitken underrelaxation to handle multiple data sets (resolves #1360 and resolves #1184), adds a preconditioner option to our Aitken acceleration and add a proper unit test for the Aitken acceleration.

## Motivation and additional information
Depends on #1373 

## Author's checklist

* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [x] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?

### CODE REVIEW

**Review by IshaanDesai — APPROVED** [review]
I tried this fix with the same case that I had mentioned in https://github.com/precice/precice/issues/1360 and it appears that the problem is fixed as the case runs further than wha it used to run.

File: src/acceleration/AitkenAcceleration.cpp [hunk 1]
```diff
@@ -55,28 +60,25 @@ void AitkenAcceleration::performAcceleration(
...
-    utils::append(values, cplData.at(id)->values());
-    utils::append(oldValues, Eigen::VectorXd(cplData.at(id)->previousIteration()));
-  }
+  concatenateCouplingData(cplData, _dataIDs, _values, _oldValues);
 
-  // Compute current residuals
-  Eigen::VectorXd residuals = values;
-  residuals -= oldValues;
+  // Compute current residual = values - oldValues
+  auto residuals = _values - _oldValues;
```
Comment (fsimonis) [inline src/acceleration/AitkenAcceleration.cpp:66]: Short info as this isn't obvious: [Eigen strongly discourages to use auto with their expressions](https://eigen.tuxfamily.org/dox/TopicPitfalls.html).

```suggestion
  Eigen::VectorXd residuals = _values - _oldValues;
```

Comment (davidscn) [reply src/acceleration/AitkenAcceleration.cpp:66]: I was not aware of that, good point, thanks!

**Review by uekerman — APPROVED** [review]
LGTM
Still, I think it would be good to add an integration test with 2 or 3 iterations to really check whether the preconditioner does the right thing if data on two different scales is used.

**Comment by BenjaminRodenberg** [comment]
@davidscn Can we merge this PR or shift it to `v3.x.x`?

**Comment by davidscn** [comment]
I don't know how outdated this is by now and adding tests requires some effort. Let's shift it.

**Comment by uekerman** [comment]
> I don't know how outdated this is by now and adding tests requires some effort. Let's shift it.

Rebasing here should be rather trivial. And the tests are nice to have, but not a strict must.
Why not merging this now? It fixes some bugs.

**Comment by davidscn** [comment]
I will merge here and add the test within the next day to not interfere with #1892

**Pull request merged**